api/src/main/java/org/apache/iceberg/metrics/ScanReport.java

```diff
@@ -0,0 +1,371 @@
+        .add("scannedDeleteManifests", scannedDeleteManifests)
+        .add("totalDataManifests", totalDataManifests)
+        .add("totalDeleteManifests", totalDeleteManifests)
+        .add("totalDataFiles", totalDataFiles)
+        .add("totalDeleteFiles", totalDeleteFiles)
+        .add("totalPlanningDuration", totalPlanningDuration)
+        .add("totalFileSizeInBytes", totalFileSizeInBytes)
+        .add("indexedDeleteFiles", indexedDeleteFiles)
+        .add("skippedDataManifests", skippedDataManifests)
+        .add("skippedDeleteManifests", skippedDeleteManifests)
+        .add("skippedDataFiles", skippedDataFiles)
+        .add("skippedDeleteFiles", skippedDeleteFiles)
+        .toString();
+  }
+
+  @SuppressWarnings("HiddenField")
+  public static class Builder {
+    private String tableName;
+    private long snapshotId = -1L;
+    private Expression filter;
+    private Schema projection;
+    private ScanMetrics scanMetrics;
+
+    private Builder() {
+    }
+
+    public Builder withTableName(String tableName) {
+      this.tableName = tableName;
+      return this;
+    }
+
+    public Builder withSnapshotId(long snapshotId) {
+      this.snapshotId = snapshotId;
+      return this;
+    }
+
+    public Builder withFilter(Expression filter) {
+      this.filter = filter;
+      return this;
+    }
+
+    public Builder withProjection(Schema projection) {
+      this.projection = projection;
+      return this;
+    }
+
+    public Builder fromScanMetrics(ScanMetrics scanMetrics) {
+      this.scanMetrics = scanMetrics;
+      return this;
+    }
+
+    public ScanReport build() {
+      Preconditions.checkArgument(null != tableName, "TableName must be non-null");
+      Preconditions.checkArgument(null != filter, "Expression filter must be non-null");
+      Preconditions.checkArgument(null != projection, "Schema projection must be non-null");
+      Preconditions.checkArgument(null != scanMetrics, "ScanMetrics must be non-null");
+      return new ScanReport(tableName, snapshotId, filter, projection,
+          scanMetrics.resultDataFiles().value(),
+          scanMetrics.totalDataManifests().value(),
+          scanMetrics.totalDeleteManifests().value(),
+          scanMetrics.scannedDataManifests().value(),
+          scanMetrics.scannedDeleteManifests().value(),
+          scanMetrics.totalDataFiles().value(),
+          scanMetrics.totalDeleteFiles().value(),
+          scanMetrics.totalPlanningDuration().totalDuration(),
+          scanMetrics.totalFileSizeInBytes().value(),
+          scanMetrics.indexedDeleteFiles().value(),
+          scanMetrics.resultDeleteFiles().value(),
+          scanMetrics.skippedDataManifests().value(),
+          scanMetrics.skippedDeleteManifests().value(),
+          scanMetrics.skippedDataFiles().value(),
+          scanMetrics.skippedDeleteFiles.value());
+    }
+  }
+
+  /**
+   * Carries all metrics for a particular scan
+   */
+  public static class ScanMetrics {
```

Comment:
I think this should be part of `ScanReport` rather than a separate class. I don't see much value in a separate class.